chia/types/eligible_coin_spends.py

```diff
@@ -27,9 +47,114 @@ class DedupCoinSpend:
+        fast_forward_spends: the fast forward spends data
+
+    Raises:
+        ValueError if none of the additions are considered to be the singleton's
+        next iteration
+    """
+    singleton_child = next(
+        (addition for addition in singleton_additions if addition.puzzle_hash == current_singleton.puzzle_hash), None
+    )
+    if singleton_child is None:
+        raise ValueError("Could not find fast forward child singleton.")
+    # Keep track of this in order to chain the next ff
+    fast_forward_spends[current_singleton.puzzle_hash] = UnspentLineageInfo(
+        coin_id=singleton_child.name(),
+        coin_amount=singleton_child.amount,
+        parent_id=singleton_child.parent_coin_info,
+        parent_amount=current_singleton.amount,
+        parent_parent_id=current_singleton.parent_coin_info,
+    )
+
+
+def perform_the_fast_forward(
+    unspent_lineage_info: UnspentLineageInfo,
+    spend_data: BundleCoinSpend,
+    fast_forward_spends: Dict[bytes32, UnspentLineageInfo],
+) -> Tuple[CoinSpend, List[Coin]]:
+    """
+    Performs a singleton fast forward, including the updating of all previous
+    additions to point to the most recent version, and updates the fast forward
+    spends data accordingly
+
+    Args:
+        unspent_lineage_info: the singleton's most recent lineage information
+        spend_data: the current spend's data
+        fast_forward_spends: the fast forward spends data
+
+    Returns:
+        CoinSpend: the new coin spend after performing the fast forward
+        List[Coin]: the updated additions that point to the new coin to spend
+
+    Raises:
+        ValueError if none of the additions are considered to be the singleton's
+        next iteration
+    """
+    new_coin = Coin(
+        unspent_lineage_info.parent_id, spend_data.coin_spend.coin.puzzle_hash, unspent_lineage_info.coin_amount
+    )
+    new_parent = Coin(
+        unspent_lineage_info.parent_parent_id,
+        spend_data.coin_spend.coin.puzzle_hash,
+        unspent_lineage_info.parent_amount,
+    )
+    # These hold because puzzle hash is not expected to change
+    assert new_coin.name() == unspent_lineage_info.coin_id
+    assert new_parent.name() == unspent_lineage_info.parent_id
+    rust_coin_spend = RustCoinSpend(
+        coin=spend_data.coin_spend.coin,
+        puzzle_reveal=RustProgram.from_bytes(bytes(spend_data.coin_spend.puzzle_reveal)),
+        solution=RustProgram.from_bytes(bytes(spend_data.coin_spend.solution)),
+    )
+    new_solution = SerializedProgram.from_bytes(
+        fast_forward_singleton(spend=rust_coin_spend, new_coin=new_coin, new_parent=new_parent)
+    )
+    singleton_child = None
+    patched_additions = []
+    for addition in spend_data.additions:
+        patched_addition = Coin(unspent_lineage_info.coin_id, addition.puzzle_hash, addition.amount)
+        patched_additions.append(patched_addition)
+        if addition.puzzle_hash == spend_data.coin_spend.coin.puzzle_hash:
```

Comment:
I think you need to check that the `amount` is odd as well

Reply:
We do these validations already on the rust side.